World_Simulation/auto_gauges.py

```diff
@@ -0,0 +1,243 @@
+import clawpack.clawutil as clawutil
+from clawpack.geoclaw import topotools
+
+##### 
+# to automatically place your gauges, first add your topography
+# and storm files below. Then, in setrun, import this module ("import auto_gauges")
+# and call "auto_gauges.return_gauge_points(rundata)" where gauges are supposed to go
+#####
+
+## directory where your data is
+DATA = os.path.join(os.environ.get('DATA_DIR', os.getcwd()))
+
+####
+# make topography file(s) into topo objects (topotools.Topography()) and
+# add to topofiles list
+####
+topofiles = []
+
+for n in [2,3,12,13,14]:
+    topo = topotools.Topography()
+    topo.read('../topo_files/world_' + str(n) + '.tt3', topo_type=3) # this is for world topo, put the path to your files
+    topofiles.append(topo)
+
+## replace with your storm
+atcf_path = os.path.join(DATA, "bal162007.dat")
+storm = Storm(path=atcf_path, file_format="ATCF")
+
+## finds the nearest points to the storm on the shoreline by using topotools 
+## makeshoreline method and finding the minimum distance from the eyelocation 
+## of the storm and the shoreline
+def points_on_shoreline(storm):
+    
+    shoreline = []
+    
+    for n in topofiles:
+        for i in n.make_shoreline_xy():
+            shoreline.append(i)
+    
+    sub_sample_storm = storm.eye_location[::3] 
+    # this takes every third coordinate, for more gauges decrease steps,
+    # for less gauges increase steps
+
+    points = np.empty([len(sub_sample_storm), 2])
+    
+    count = 0
+    
+    for n in sub_sample_storm:
+    
+        nearest_shore = near_shoreline(n[0], n[1], shoreline)
+        
+        points[count] = (shoreline[nearest_shore])
+        count = count + 1
+        
+    
+    return points
+
+## if storm file is read in as geoclaw file, comment out previous method
+## and uncomment this method:
+
+# def points_on_shoreline(storm):
+    
+#     location_x = storm.eye_location[0]
+    
+#     location_y = storm.eye_location[1]
+    
+#     shoreline = []
+    
+#     for n in topofiles:
+#         for i in n.make_shoreline_xy():
+#             shoreline.append(i)
+
+#     points = np.empty([len(location_y), 2])
+    
+#     count = 0
+    
+#     for n in range(0, len(points)):
+    
+#         nearest_shore = near_shoreline(location_x[n], location_y[n], shoreline)
+        
+#         points[count] = (shoreline[nearest_shore])
+#         count = count + 1
+        
+    
+#     return points
+
+
+## returns shoreline closest to storm by calculating distance
+## between storm coordinate and every point on shoreline
+def near_shoreline(storm_x, storm_y, shoreline):
+    
+    dis = []
+    
+    for n in shoreline:
+        dis.append((distance(storm_x, storm_y, n[0], n[1])))
+        
+    minimum = dis.index(min(dis))
+    return minimum
+ 
+    
+## find distance between two points by using haversine formula    
+def distance(x1, y1, x2, y2):
+    
+    # convert to radians
+    x1 = x1*math.pi/180
```

Comment:
Probably want to use `numpy.pi` rather than import `math`'s version since we already need `numpy`.